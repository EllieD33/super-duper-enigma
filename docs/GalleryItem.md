# GalleryItem Component

## Description

The `GalleryItem` component renders an image inside a container, with optional overlay text. It supports different sizes and displays a loading skeleton or fallback image in case of errors or while loading.

## Props

### `GalleryItemProps`

| Prop             | Type                             | Required | Description                                                                                               |
| ---------------- | -------------------------------- | -------- | --------------------------------------------------------------------------------------------------------- |
| `imageUrl`       | `string`                         | ✅       | The URL of the image to display.                                                                          |
| `altText`        | `string`                         | ✅       | The alt text for the image.                                                                               |
| `imageId`        | `number`                         | ✅       | Unique identifier for the image.                                                                          |
| `size`           | `"small" \| "medium" \| "large"` | ❌       | Optional size of the image. Defaults to `"medium"`.                                                       |
| `overlayText`    | `string`                         | ❌       | Optional text to display over the image as an overlay.                                                    |
| `overlayDisplay` | `"hover" \| "static"`            | ❌       | Specifies when the overlay text should appear: `"hover"` for hover effect, `"static"` for always visible. |

## Behaviour

- The `GalleryItem` component renders an image inside a container.
- If the image is still loading, a skeleton loader will be displayed.
- If the image fails to load, a fallback image will be shown instead.
- Optional overlay text can be displayed, either always visible (`"static"`) or shown only on hover (`"hover"`).
- Supports multiple sizes of images: small, medium, and large.
- The component uses a lazy loading strategy for images to improve performance.

[Explore GalleryItem in Storybook](http://localhost:6006/?path=/docs/library-galleries-galleryitem--docs&args=error:&globals=viewport:largeMobile)

## Notes

- **Fallback Image**: If the `imageUrl` fails to load or has an error, a placeholder image (`https://placehold.co/600`) is shown instead.
- **Overlay Display**: When `overlayDisplay` is set to `"hover"`, the overlay text appears only when the user hovers over the image. If set to `"static"`, the text remains visible at all times.
- The **loading skeleton** is shown while the image is still loading, providing a better user experience during image load times.
- The `GalleryItem` component has a flexible design, adapting to different image sizes depending on the `size` prop.
