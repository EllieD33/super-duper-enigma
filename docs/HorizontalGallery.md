# HorizontalGallery Component

## Description

The `HorizontalGallery` component renders a horizontally scrollable gallery of images. It uses the `GalleryItem` component to display each image, with optional overlay text and a hover or static display mode. It is perfect for showcasing a collection of images in a scrollable container.

## Props

### `HorizontalGalleryProps`

| Prop     | Type                 | Required | Description                                                                                                    |
| -------- | -------------------- | -------- | -------------------------------------------------------------------------------------------------------------- |
| `images` | `GalleryItemProps[]` | âœ…       | An array of images to be displayed in the gallery. Each image must adhere to the `GalleryItemProps` interface. |

### `GalleryItemProps`

See [GalleryItem Documentation](/docs/GalleryItem.md)

## Behaviour

- The `HorizontalGallery` component displays a horizontally scrollable gallery of images.
- Each image is rendered using the `GalleryItem` component.
- Optionally, an overlay text can be displayed on top of the image. The display mode of the overlay can be controlled with the `overlayDisplay` prop (`"hover"` or `"static"`).
- The gallery is wrapped in the `Scrollbar` component to make it scrollable.

[Explore HorizontalGallery in Storybook](http://localhost:6006/?path=/story/library-galleries-horizontalgallery--default&globals=viewport:largeMobile)

## Notes

- The `GalleryItem` component is used to display each individual image.
- The `Scrollbar` component provides the horizontal scrolling behavior.
