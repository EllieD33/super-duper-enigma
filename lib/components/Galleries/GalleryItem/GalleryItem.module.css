.image {
    border-radius: 15px;
    object-fit: cover;
}

.small {
    height: 100px;
    width: 100px;
}

.medium {
    height: 200px;
    width: 200px;
}

.large {
    height: 400px;
    width: 400px;
}

.skeletonItem {
    position: relative;
    background-color: #f0f0f0;
    overflow: hidden;
    border-radius: 15px;
    display: flex;
    justify-content: center;
    align-items: center;
}

.skeletonItem::after {
    content: "";
    position: absolute;
    top: 0;
    left: -100%;
    width: 100%;
    height: 100%;
    background: linear-gradient(
        to right,
        #f0f0f0 0%,
        #e0e0e0 20%,
        #f0f0f0 40%,
        #f0f0f0 100%
    );
    animation: shimmer 1.5s infinite linear;
}

@keyframes shimmer {
    0% {
        left: -100%;
    }
    100% {
        left: 100%;
    }
}
