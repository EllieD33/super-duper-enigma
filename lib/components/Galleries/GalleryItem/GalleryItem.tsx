import React, { ReactElement, useState } from "react";
import styles from "./GalleryItem.module.css";
import clsx from "clsx";

export interface GalleryItemProps {
    imageUrl: string;
    altText: string;
    imageId: number;
    size?: "small" | "medium" | "large";
}

const GalleryItem = ({
    imageUrl,
    altText,
    imageId,
    size = "medium",
}: GalleryItemProps): ReactElement => {
    const [isLoading, setIsLoading] = useState<boolean>(true);
    const [hasError, setHasError] = useState<boolean>(false);

    const galleryItemStyles = clsx(styles.image, styles[size]);
    const skeletonItemStyles = clsx(styles.SkeletonItem, styles[size]);
    const fallbackImage = "https://placehold.co/600";

    const handleImageLoad = () => {
        setIsLoading(false);
    };

    const handleImageError = () => {
        setIsLoading(false);
        setHasError(true);
    };

    return (
        <div id={`gallery-item-${imageId}`}>
            <img
                className={galleryItemStyles}
                src={hasError ? fallbackImage : imageUrl}
                alt={hasError ? "image not found" : altText}
                loading="lazy"
                onLoad={handleImageLoad}
                onError={handleImageError}
                data-testid="galleryItemImage"
            />
            {isLoading && !hasError && (
                <div
                    className={skeletonItemStyles}
                    data-testid="skeleton"
                ></div>
            )}
        </div>
    );
};

export default GalleryItem;
