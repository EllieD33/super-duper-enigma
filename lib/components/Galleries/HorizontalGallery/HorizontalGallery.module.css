.container {
    display: flex;
}

.small {
    height: 100px;
}

.medium {
    height: 200px;
}

.large {
    height: 400px;
}
