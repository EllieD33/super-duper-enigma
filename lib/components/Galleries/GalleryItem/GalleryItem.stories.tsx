import type { Meta, StoryObj } from "@storybook/react";
import GalleryItem, { GalleryItemProps } from "./GalleryItem";

const meta = {
    title: "Library/Galleries/GalleryItem",
    component: GalleryItem,
    parameters: {
        layout: "centered",
    },
    tags: ["autodocs"],
    argTypes: {},
} satisfies Meta<typeof GalleryItem>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        imageId: 1,
        imageUrl: "https://cdn.pixabay.com/photo/2023/08/08/15/01/flower-8177578_640.jpg",
        altText: "white flowers wth dark green leaves behind them",
    } as GalleryItemProps,
};
