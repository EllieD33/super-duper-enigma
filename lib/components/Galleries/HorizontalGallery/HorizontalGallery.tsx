import { ReactElement } from "react";
import GalleryItem, { GalleryItemProps } from "../GalleryItem/GalleryItem";
import clsx from "clsx";
import styles from "./HorizontalGallery.module.css"

export interface HorizontalGalleryProps {
    images: GalleryItemProps[];
    size?: 'small' | 'medium' | 'large'
}

const HorizontalGallery = ({
    images,
    size = 'medium'
}: HorizontalGalleryProps): ReactElement => {
    const galleryStyles = clsx(styles.container, styles[size])
    return (
            <div className={galleryStyles}>
                {images &&
                    images.map((image) => (
                        <div key={image.imageId}>
                            <GalleryItem {...image} />
                        </div>
                    ))}
            </div>
    );
};

export default HorizontalGallery;
