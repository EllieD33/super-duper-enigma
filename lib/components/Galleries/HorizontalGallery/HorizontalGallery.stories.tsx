import type { Meta, StoryObj } from "@storybook/react";
import HorizontalGallery, {
    HorizontalGalleryProps,
} from "././HorizontalGallery";

const meta = {
    title: "Library/Galleries/HorizontalGallery",
    component: HorizontalGallery,
    parameters: {
        layout: "centered",
    },
    tags: ["autodocs"],
    argTypes: {},
} satisfies Meta<typeof HorizontalGallery>;

export default meta;
type Story = StoryObj<typeof meta>;

const imageArray = [
    {
        imageId: 1,
        imageUrl: "https://cdn.pixabay.com/photo/2023/08/08/15/01/flower-8177578_640.jpg",
        altText: "image alt text",
    },
    {
        imageId: 2,
        imageUrl: "https://cdn.pixabay.com/photo/2023/04/21/11/51/flower-7941764_640.jpg",
        altText: "image alt text",
    },
    {
        imageId: 3,
        imageUrl: "https://cdn.pixabay.com/photo/2023/06/04/12/00/rose-8039502_640.jpg",
        altText: "image alt text",
    },
    {
        imageId: 4,
        imageUrl: "https://cdn.pixabay.com/photo/2023/04/08/18/01/flower-7909902_640.jpg",
        altText: "image alt text",
    },
    {
        imageId: 5,
        imageUrl: "https://cdn.pixabay.com/photo/2018/02/09/21/46/rose-3142529_640.jpg",
        altText: "image alt text",
    },
    {
        imageId: 6,
        imageUrl: "https://cdn.pixabay.com/photo/2016/07/11/21/23/water-lily-1510707_640.jpg",
        altText: "image alt text",
    },
    {
        imageId: 7,
        imageUrl: "https://cdn.pixabay.com/photo/2019/05/14/16/43/flower-4202825_640.jpg",
        altText: "image alt text",
    },
    {
        imageId: 8,
        imageUrl: "https://cdn.pixabay.com/photo/2023/07/28/20/36/flower-8155951_640.jpg",
        altText: "image alt text",
    }
];


export const Default: Story = {
    args: {
        images: imageArray,
    } as HorizontalGalleryProps,
};
