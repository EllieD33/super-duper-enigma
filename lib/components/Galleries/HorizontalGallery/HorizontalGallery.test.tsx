import React from "react";
import { render, screen } from "@testing-library/react";
import HorizontalGallery, { HorizontalGalleryProps } from "./HorizontalGallery";
import mockImages from "./testData/testData";
import { GalleryItemProps } from "../GalleryItem/GalleryItem";

const defaultProps: HorizontalGalleryProps = {
    images: mockImages,
};

describe("HorizontalGallery", () => {
    it("should render the default component", () => {
        const { container } = render(<HorizontalGallery {...defaultProps} />);
        expect(container).toMatchSnapshot();
    });

    it("should render the correct number of gallery items", () => {
        const { getAllByAltText } = render(
            <HorizontalGallery {...defaultProps} />
        );
        expect(getAllByAltText(/image/)).toHaveLength(mockImages.length);
    });

    it("should pass the correct props to GalleryItem", () => {
        const { getByAltText } = render(
            <HorizontalGallery {...defaultProps} />
        );
        mockImages.forEach((image: GalleryItemProps) => {
            const imgElement = getByAltText(image.altText);
            expect(imgElement).toHaveAttribute("src", image.imageUrl);
        });
    });

    it("should apply the correct container class", () => {
        const { container } = render(<HorizontalGallery {...defaultProps} />);
        expect(container.firstChild).toHaveClass("container");
    });

    it("should render nothing when images array is empty", () => {
        render(<HorizontalGallery images={[]} />);
        const container = screen.queryByTestId("container");
        expect(container).not.toBeInTheDocument();
    });
});
