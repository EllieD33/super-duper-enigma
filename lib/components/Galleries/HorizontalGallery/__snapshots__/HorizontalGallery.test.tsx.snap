// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalGallery should render the default component 1`] = `
<div>
  <div
    class="container"
    data-testid="gallery-container"
  >
    <div>
      <div
        id="gallery-item-1"
      >
        <img
          alt="image1 alt text"
          class="image medium"
          data-testid="galleryItemImage"
          loading="lazy"
          src="https://cdn.pixabay.com/photo/2023/08/08/15/01/flower-8177578_640.jpg"
        />
        <div
          class="SkeletonItem medium"
          data-testid="skeleton"
        />
      </div>
    </div>
    <div>
      <div
        id="gallery-item-2"
      >
        <img
          alt="image2 alt text"
          class="image medium"
          data-testid="galleryItemImage"
          loading="lazy"
          src="https://cdn.pixabay.com/photo/2023/04/21/11/51/flower-7941764_640.jpg"
        />
        <div
          class="SkeletonItem medium"
          data-testid="skeleton"
        />
      </div>
    </div>
    <div>
      <div
        id="gallery-item-3"
      >
        <img
          alt="image3 alt text"
          class="image medium"
          data-testid="galleryItemImage"
          loading="lazy"
          src="https://cdn.pixabay.com/photo/2023/06/04/12/00/rose-8163000_640.jpg"
        />
        <div
          class="SkeletonItem medium"
          data-testid="skeleton"
        />
      </div>
    </div>
  </div>
</div>
`;
